src/data-model-providers/codedriven/CodeDrivenDataModelProvider.cpp

```diff
@@ -0,0 +1,337 @@
+        auto serverClusters = serverClusterBuilder.TakeBuffer();
+
+        for (auto * serverCluster : serverClusters)
+
+        {
+            CHIP_ERROR err = serverCluster->Startup(*mServerClusterContext);
+            if (err != CHIP_NO_ERROR)
+            {
+                had_failure = true;
+                VerifyOrReturnError(!serverCluster->GetPaths().empty(), CHIP_ERROR_INCORRECT_STATE);
+                // Log startup failure for the specific cluster
+                const ConcreteClusterPath path = serverCluster->GetPaths().front();
+                ChipLogError(DataManagement, "Cluster " ChipLogFormatMEI " on Endpoint %u startup failed: %" CHIP_ERROR_FORMAT,
+                             ChipLogValueMEI(path.mClusterId), path.mEndpointId, err.Format());
+            }
+        }
+    }
+    if (had_failure)
+    {
+        return CHIP_ERROR_HAD_FAILURES;
+    }
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR CodeDrivenDataModelProvider::Shutdown()
+{
+    // Shutdown all server clusters across all endpoints
+    for (auto * endpointProvider : mEndpointProviderRegistry)
+    {
+        ReadOnlyBufferBuilder<ServerClusterInterface *> serverClusterBuilder;
+        ReturnErrorOnFailure(endpointProvider->ServerClusterInterfaces(serverClusterBuilder));
+        auto serverClusters = serverClusterBuilder.TakeBuffer();
+
+        for (auto * serverCluster : serverClusters)
+        {
+            // TODO: The API for serverCluster->Shutdown is swallowing errors here. We should update it to return CHIP_ERROR.
+            serverCluster->Shutdown();
+        }
+    }
+    mServerClusterContext.reset();
+    return CHIP_NO_ERROR;
+}
+
+DataModel::ActionReturnStatus CodeDrivenDataModelProvider::ReadAttribute(const DataModel::ReadAttributeRequest & request,
+                                                                         AttributeValueEncoder & encoder)
+{
+    EndpointProviderInterface * epProvider = GetEndpointProvider(request.path.mEndpointId);
+    if (epProvider == nullptr)
+    {
+        return Status::UnsupportedEndpoint;
+    }
+    ServerClusterInterface * serverCluster = epProvider->GetServerCluster(request.path.mClusterId);
+    if (serverCluster == nullptr)
+    {
+        return Status::UnsupportedCluster;
+    }
+    return serverCluster->ReadAttribute(request, encoder);
+}
+
+DataModel::ActionReturnStatus CodeDrivenDataModelProvider::WriteAttribute(const DataModel::WriteAttributeRequest & request,
+                                                                          AttributeValueDecoder & decoder)
+{
+    EndpointProviderInterface * epProvider = GetEndpointProvider(request.path.mEndpointId);
+    if (epProvider == nullptr)
+    {
+        return Status::UnsupportedEndpoint;
+    }
+    ServerClusterInterface * serverCluster = epProvider->GetServerCluster(request.path.mClusterId);
+    if (serverCluster == nullptr)
+    {
+        return Status::UnsupportedCluster;
+    }
```

Comment:
the "get the server interface" seems very commonly used, so I wonder if ther would be a way to re-use this part if it saves flash? Could be a TODO in the future when we actually have flash statistics.